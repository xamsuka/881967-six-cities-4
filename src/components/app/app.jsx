import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Route, Router, Switch, Redirect} from 'react-router-dom';
import Main from '../main/main.jsx';
import SignIn from '../sign-in/sign-in.jsx';
import PrivateRoute from '../private-route/private-route.jsx';
import Favorites from '../favorites/favorites.jsx';
import DetailedOffer from '../detailed-offer/detailed-offer.jsx';
import {ActionCreator as ActionCreatorApplication} from '../../reducers/application/reducer.js';
import {getSortedOffers} from '../../reducers/data/selectors.js';
import {getStatusLoadingOffers} from '../../reducers/application/selectors.js';
import {getCurrentCity} from '../../reducers/application/selectors.js';
import {Operations as UserOperations} from '../../reducers/user/reducer.js';
import {getUserData} from '../../reducers/user/selectors.js';
import {createBrowserHistory} from 'history';

export const history = createBrowserHistory();

class App extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {
      offers,
      currentCity,
      onChangeCurrentCity,
      onLogin,
      userData,
      isLoading,
    } = this.props;

    return (
      <Router history={history}>
        <Switch>
          <Route exact path="/">
            <Main offers = {offers} currentCity = {currentCity} onChangeCurrentCity = {onChangeCurrentCity} isLoading = {isLoading} userData = {userData} />;
          </Route>
          <Route exact path='/offer/:id' component={DetailedOffer}/>
          <Route exact path='/login' >
            {Object.keys(userData).length ? <Redirect to="/" /> : <SignIn onSubmit = {onLogin} history={history} /> }
          </Route>
          <PrivateRoute
            exact
            path='/favorites'
            render={() => {
              return (
                <Favorites
                  userData = {userData}
                />
              );
            }}
          />
        </Switch>
      </Router>
    );
  }
}

App.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    city: PropTypes.shape({
      location: PropTypes.shape({
        latitude: PropTypes.number.isRequired,
        longitude: PropTypes.number.isRequired,
        zoom: PropTypes.number.isRequired,
      }).isRequired,
      name: PropTypes.string.isRequired,
    }),
    previewPhoto: PropTypes.string.isRequired,
    photos: PropTypes.array.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    isPremium: PropTypes.bool.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    type: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    countDedrooms: PropTypes.number.isRequired,
    maxGuests: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    features: PropTypes.array.isRequired,
    infoOwner: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string.isRequired,
      isPro: PropTypes.bool.isRequired,
    }),
    coords: PropTypes.shape({
      latitude: PropTypes.number.isRequired,
      longitude: PropTypes.number.isRequired,
      zoom: PropTypes.number.isRequired,
    }).isRequired,
  })).isRequired,
  currentCity: PropTypes.oneOf([`Paris`, `Cologne`, `Brussels`, `Amsterdam`, `Hamburg`, `Dusseldorf`]).isRequired,
  onChangeCurrentCity: PropTypes.func.isRequired,
  onLogin: PropTypes.func.isRequired,
  userData: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    avatarUrl: PropTypes.string,
    isPro: PropTypes.bool,
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  offers: getSortedOffers(state),
  currentCity: getCurrentCity(state),
  isLoading: getStatusLoadingOffers(state),
  userData: getUserData(state),
});

const mapDispatchToProps = (dispatch) => ({
  onChangeCurrentCity: (evt) => {
    evt.preventDefault();
    dispatch(ActionCreatorApplication.changeCities(evt));
  },
  onLogin: (userData) => {
    dispatch(UserOperations.authorizeUser(userData));
  },
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
