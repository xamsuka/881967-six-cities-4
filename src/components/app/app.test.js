import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {App} from './app.jsx';

const mockStore = configureStore([]);

const offers = [
  {
    id: 1,
    city: {
      location: {
        latitude: 52.374,
        longitude: 4.897976,
        zoom: 13,
      },
      name: `Paris`
    },
    previewPhoto: `https://canadskaya-izba.ru/img/doma/karkas2.jpg`,
    photos: [`https://canadskaya-izba.ru/img/doma/karkas2.jpg`, `https://www.krasdom.com/galfotobig/584.jpg`, `https://www.krasdom.com/galfotobig/561.jpg`, `https://www.artis21.ru/upload/iblock/f16/f164e862e34e212180e03f76c9226ee7.jpg`],
    title: `Очень красивый и уютный дом`,
    description: `Описание у всех одинаковое, потому что лень придумывать.`,
    isPremium: false,
    isFavorite: false,
    type: `Apartment`,
    rating: 4,
    countDedrooms: 3,
    maxGuests: 3,
    price: 500,
    features: [`Wifi`, `Heating`, `Kitchen`, `Cable TV`, `Washing machine`],
    infoOwner: {
      id: 1,
      name: `Vladislav Kozlovsky`,
      avatar: `https://api.adorable.io/avatars/285/abott@adorable.png`,
      isPro: true,
    },
    coords: {
      latitude: 52.39,
      longitude: 4.8,
      zoom: 16,
    },
  },
  {
    id: 2,
    city: {
      location: {
        latitude: 52.374,
        longitude: 4.897976,
        zoom: 12,
      },
      name: `Paris`
    },
    previewPhoto: `https://canadskaya-izba.ru/img/doma/karkas2.jpg`,
    photos: [`https://canadskaya-izba.ru/img/doma/karkas2.jpg`, `https://www.krasdom.com/galfotobig/584.jpg`, `https://www.krasdom.com/galfotobig/561.jpg`, `https://www.artis21.ru/upload/iblock/f16/f164e862e34e212180e03f76c9226ee7.jpg`],
    title: `Просто хорошее место`,
    description: `Описание у всех одинаковое, потому что лень придумывать.`,
    isPremium: true,
    isFavorite: false,
    type: `Apartment`,
    rating: 5,
    countDedrooms: 4,
    maxGuests: 2,
    price: 200,
    features: [`Wifi`, `Heating`, `Kitchen`, `Cable TV`, `Washing machine`],
    infoOwner: {
      id: 2,
      name: `Vladislav Petrov`,
      avatar: `https://api.adorable.io/avatars/285/abott@adorable.png`,
      isPro: false,
    },
    coords: {
      latitude: 52.39,
      longitude: 4.8,
      zoom: 16,
    },
  },
];

test(`App render`, () => {
  const store = mockStore({
    "APPLICATION": {
      city: `Paris`,
      currentSort: `Popular`,
    },
    "DATA": {
      offers
    },
    "USER": {
      authorizationStatus: `USER_AUTH`,
    }
  });

  const app = renderer
    .create(
        <Provider store={store}>
          <App
            offers = {offers}
            currentCity = {`Paris`}
            onChangeCurrentCity = {() => {}}
            onLogin = {() => {}}
            userData = {{}}
            isLoading = {false} />
        </Provider>
        , {
          createNodeMock: () => {
            return {};
          }}).toJSON();

  expect(app).toMatchSnapshot();

});
