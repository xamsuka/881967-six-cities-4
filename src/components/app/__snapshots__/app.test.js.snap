// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render 1`] = `ReactWrapper {}`;
